aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -571,48 +566,21 @@ Tensor& _sparse_csr_mm_out(
 }
 
 Tensor _sparse_csr_mm(const Tensor& mat1, const Tensor& mat2) {
-  if (mat1.is_sparse_csr() && mat2.is_sparse_csr()) {
-    // Return sparse
-    // TODO: replace with at::zeros when it's implemented for sparse csr
-    return at::addmm(
-        at::empty({mat1.size(0), mat2.size(1)}, mat2.options()),
-        mat1,
-        mat2,
-        0.0,
-        1.0);
-  }
-  if ((mat1.layout() == kSparseCsc || mat1.layout() == kSparseCsr) &&
-      (mat2.layout() == kSparseCsc || mat2.layout() == kSparseCsr)) {
-    // TODO: Expensive conversion to CSR. Should add native support for CSC.
-    // Covers CSC @ CSR
-    // Covers CSR @ CSC
-    // Covers CSC @ CSC
-    return _sparse_csr_mm(mat1.to_sparse_csr(), mat2.to_sparse_csr());
-  }
-  if (mat1.layout() == kSparseCsc && mat2.layout() == c10::kStrided) {
-    // TODO: This is a costly conversion. We should have
-    // native support for CSC.
-    return _sparse_csr_mm(mat1.to_sparse_csr(), mat2);
-  }
-  if (mat1.is_sparse_csr() && mat2.layout() == c10::kStrided) {
-    // Return dense
-    return at::addmm(
-        at::zeros({mat1.size(0), mat2.size(1)}, mat2.options()),
-        mat1,
-        mat2,
-        0.0,
-        1.0);
-  }
-  if (mat1.layout() == c10::kStrided && mat2.is_sparse_csr()) {
-    // Return dense
-    return at::addmm(
-        at::zeros({mat1.size(0), mat2.size(1)}, mat1.options()),
-        mat1,
-        mat2,
-        0.0,
-        1.0);
+  Tensor result;
+  // If either input is strided result will be strided
+  if (mat1.layout() == kStrided) {
+    result = at::zeros({mat1.size(-2), mat2.size(-1)}, mat1.options());
+  } else if (mat2.layout() == kStrided) {
+    result = at::zeros({mat1.size(-2), mat2.size(-1)}, mat2.options());
+  } else {
+    // Otherwise the result is sparse, result layout must be csr for resize of
+    // output to mat1/mat2 will be coerced to mm compatible layout if possible
```

Comment:
I guess one reason for `CSR @ CSR -> CSR` would be that we have `COO @ COO -> COO` but that would be still insufficient to require, say, `CSC @ CSC -> CSR`, imho.

So, a possible rule for `mm` output layout could be "mm(mat1, mat2).layout == mat1.layout".

Reply:
FWIW I agree that as a default `sparese_compressed @ spares_compressed -> strided` does make the most sense, but there is only support for it on CPU.  

On mkl we have the `addmm_sparse_input_dense_result` variant, there is no similar api available on cuda. So if we initialize result here with `strided` layout, when we eventually call into `at::native::sparse::impl::cuda::addmm_out_sparse_csr(` there will never be a combination caught with `mat1(sparse), mat2(sparse), result(strided)`

The only way to do `torch.mm(a, b)` on cuda (with a/b sparse_csr) would be to pre allocate the result and call addmm directly.

With the current choice of `sparse_csr` as the result layout we have the nice property that the supported combinations for mat1/mat2 layouts is exactly the same. 

> So So, a possible rule for `mm` output layout could be "mm(mat1, mat2).layout == mat1.layout"

That would also not be ideal, as we cannot evaluate `sparse @ strided -> sparse` on any backend.

I do see what you are getting at and it would be nice to have a consistent treatment with some type of rule, but right now we have too many holes in the layout coverage. 

IMO right now it makes the most sense to select the result layout which gives the most supported combinations on CUDA, and will consistently work across cpu/gpu.

Reply:
I'd like to think that any lack of support for a given input/output layout combination to `mm` is a matter of implementing the support. When to do that depends on the need and prioritization.

My main concern is that if we design sparse `mm` support based on the available implementations, we may be digging a BC hole where we will fall sooner or later, and getting out of it will be much harder if impossible. Leaving certain combinations unimplemented or providing inefficient implementations instead (as is the `to_sparse_csr/transpose` usage tricks), but having consistent semantics that is an implementation detail independent, is a more reasonable approach, IMHO.

Reply:
@cpuhrsch care to weigh in on this?

Reply:
Let's prioritize CUDA. Let's make sure our behavior is device independent even if it requires additional work on CPU. 

The current high priority item is linear(input: 3D dense, weight: 2D sparse, bias: 1D dense) -> 3D dense. This should not involve any mm(sparse, sparse).

If a combination of inputs is not supported, let's throw an exception and tell the user. The user can always convert back to dense before calling an operation, albeit they'll miss out on the performance gains.

Reply:
See the latest update: 

The result will be strided if either input is strided. Otherwise the result will have the layout of mat1.  Some of the early layout conversions had to be left in here to preserve the existing behavior. They can be removed later when working on sparse x sparse mm which is OT for this PR.